cmd/cosign/main.go

```diff
@@ -47,22 +49,39 @@ func main() {
 	}
 
 	if err := root.Parse(os.Args[1:]); err != nil {
-		if *verbose {
-			fmt.Print("verbose!")
+		printErrAndExit(err)
+	}
+
+	if *outputFilename != "" {
+		out, err := os.Create(*outputFilename)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "error: %v\n", errors.Wrapf(err, "Error creating output file %s", *outputFilename))
+			os.Exit(1)
 		}
-		fmt.Fprintf(os.Stderr, "error: %v\n", err)
-		os.Exit(1)
+		stdout := os.Stdout
+		stderr := os.Stderr
+		defer func() {
+			os.Stdout = stdout
+			os.Stderr = stderr
```

Comment:
Do you think it makes sense to always capture both stdout and stderr? In general I've tried to be careful and put actual "output" in stdout, and user-centered "messages" to stderr.

Reply:
I'm not sure I'm clear on the distinction. For example, https://github.com/SigStore/cosign/blob/2befb2893ee5ac5e697ed09fd4c146eef98a6ab7/cmd/cosign/cli/verify.go#L127 contains print statements to `stderr` for results of verifying claims, tlog entry, etc., while logging error messages to `stdout`. I think I would expect the opposite in that case and so I wasn't able to find a consistent pattern throughout to understand a distinction. As a result I found needing to see both stdout and stderr messages. But that's not necessarily a reason to log both to a file, and in principle I agree and think it's useful to separate them or at least duplicate them if nothing else.

We can just log to `stdout` for now (leave `stderr` untouched), file an issue to make `stdout`/`stderr` logging convention consistent and perhaps we'll need to consider revamping the entire print approach to use a more abstracted logging mechanism. WDYT?

Reply:
> We can just log to `stdout` for now (leave `stderr` untouched), file an issue to make `stdout`/`stderr` logging convention consistent and perhaps we'll need to consider revamping the entire print approach to use a more abstracted logging mechanism. WDYT?

Perfect! I *think* that code you linked is actually working as I intended it to, but I admit the intention isn't very clear and could probably use improvements.

Reply:
Cool! Filed https://github.com/sigstore/cosign/issues/263.